### DISCUSSION TITLE
$HOME installation of brew 4.0.0+ shows no updates unless disabling API; unsupported or bug?

### DISCUSSION THREAD

**Discussion by itaiferber** [Homebrew/discussions#4257] [Everyday usage] [answered]
### Output of `brew config`

```shell
HOMEBREW_VERSION: 4.0.3-18-g059945a
ORIGIN: https://github.com/Homebrew/brew
HEAD: 059945ae0d618d43e2412a5ec8d63dedab97a8fb
Last commit: 3 hours ago
Core tap origin: https://github.com/Homebrew/homebrew-core
Core tap HEAD: 9f6f46344e65033096b2ba68001ec38467e84a8f
Core tap last commit: 39 minutes ago
Core tap branch: master
Core tap JSON: 21 Feb 22:29 UTC
HOMEBREW_PREFIX: /Users/itai/Homebrew
HOMEBREW_REPOSITORY: /Users/itai/Homebrew
HOMEBREW_CELLAR: /Users/itai/Homebrew/Cellar
HOMEBREW_CASK_OPTS: []
HOMEBREW_EDITOR: nvim
HOMEBREW_MAKE_JOBS: 10
Homebrew Ruby: 2.6.10 => /System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/bin/ruby
CPU: 10-core 64-bit arm_firestorm_icestorm
Clang: 14.0.0 build 1400
Git: 2.39.1 => /Users/itai/Homebrew/bin/git
Curl: 7.85.0 => /usr/bin/curl
macOS: 13.1-arm64
CLT: 14.2.0.0.1.1668646533
Xcode: 14.2
Rosetta 2: false
```

### Output of `brew doctor`

```shell
Please note that these warnings are just used to help the Homebrew maintainers
with debugging if you file an issue. If everything you use Homebrew for is
working fine: please don't worry or file an issue; just ignore this. Thanks!

Warning: Your Homebrew's prefix is not /opt/homebrew.

Many of Homebrew's bottles (binary packages) can only be used with the default prefix.
Consider uninstalling Homebrew and reinstalling into the default prefix.
It is expected behaviour that some formulae will fail to build in this unsupported configuration.
It is expected behaviour that Homebrew will be buggy and slow.
Do not create any issues about this on Homebrew's GitHub repositories.
Do not create any issues even if you think this message is unrelated.
Any opened issues will be immediately closed without response.
Do not ask for help from Homebrew or its maintainers on social media.
You may ask for help in Homebrew's discussions but are unlikely to receive a response.
Try to figure out the problem yourself and submit a fix as a pull request.
We will review it but may or may not accept it.
```

### Description of issue

Since upgrading to `brew` 4.0.0, running `brew update [--force]` has shown no updates available, and no outdated formulae, only updating `brew` itself. Attempting to upgrade either formulae or casks with `brew upgrade --force` has also failed, with the claim that the latest version is already installed.

However, I just ran `env HOMEBREW_NO_INSTALL_FROM_API=true brew update`, which has updated my local taps (just `homebrew-core` and `homebrew-cask`) and revealed 39 outdated formulae and 5 outdated casks. I was able to upgrade just fine with `env HOMEBREW_NO_INSTALL_FROM_API=true brew upgrade`.

I'm wondering:

1. Is this expected behavior? (Because I still have local taps of the formulae repos, or for some other reason?)
2. Is this somehow related to installing Homebrew outside of the default prefix? (Since the last time I checked, I see that this configuration has dropped from "not recommended" to explicitly ["unsupported"](https://docs.brew.sh/Installation#untar-anywhere-unsupported), which is somewhat disheartening)
3. Is there a way to fix installation from the API, or will I need to permanently set this env var?

Searching around doesn't reveal anyone else obviously having this problem, but I may have missed something! Thanks for any help.

#### Update

It does appear that after the initial runs above with `HOMEBREW_NO_INSTALL_FROM_API=true`, I am now again seeing regular updates and am able to upgrade without explicitly running with the env var set. I don't know if this is expected either, but this does at least allow me to return to my routine without needing to worry about the env var.

**Comment by Bo98** [comment, marked as answer]
This indeed looks like a bug, I'll take a look.

**Reply by itaiferber** [reply]
Thanks, Bo! Much appreciated!

**Reply by itaiferber** [reply]
It does appear, by the way, that after this initial update/upgrade cycle with `HOMEBREW_NO_INSTALL_FROM_API` enabled, I'm now getting regular updates again without the flag set.

**Reply by Bo98** [reply]
I've pushed a change that should fix this issue: https://github.com/Homebrew/brew/commit/e86a8c2477032bda22c075f948f2c526c7ce4e02

**Reply by itaiferber** [reply]
Thanks, @Bo98! The changes in that PR look great (at least to someone who isn't very familiar with the Homebrew codebase), and I can confirm that after upgrading to `4.0.4`, things continue to look good. Much appreciated!
